File: 20ValidParentheses.java
import java.util.*;

public class ValidParentheses {
  public static void main(String[] args) {
    String s = "()";
    System.out.println(isValid(s));
  }

  public static boolean isValid(String s) {
    if (s == null || s.isEmpty())
        return true;

    Map<Character, Character> map = new HashMap<>();
    map.put('(', ')');
    map.put('{', '}');
    map.put('[', ']');

    Stack<Character> stack = new Stack<>();

    for (char c : s.toCharArray()) {

        if (c == '[' || c == '(' || c == '{')
            stack.push(c);
        else {
            if (stack.isEmpty() || map.get(stack.peek()) != c)
                return false;
            stack.pop();
        }
    }

    return stack.isEmpty();
}

}

File: 20ValidParentheses.go
package main

import "fmt"

func main() {
	s := "{}"
	fmt.Print(isValid(s))
}

var m = map[rune]rune{
	'(': ')',
	'[': ']',
	'{': '}',
}

func isValid(s string) bool {
	stack := []rune{}
	for _, char := range s {
		if _, ok := m[char]; ok {
			stack = append(stack, char)
		} else if len(stack) == 0 || m[stack[len(stack)-1]] != char {
			return false
		} else {
			stack = stack[:len(stack)-1]
		}
	}

	return len(stack) == 0
}
